code/data_structures/src/heaps/binomial_heap/binomial_heap.hpp

```diff
@@ -0,0 +1,609 @@
+binomialNode<T>::~binomialNode()
+{
+	//std::cout << "\nDeleting " << value;
+
+	if(child)
+		delete child;
+
+	if(sibling)
+		delete sibling;
+
+}
+
+template<typename T>
+int binomialNode<T>::size(int & i)
+{
+	i ++;
+	if(sibling)
+		sibling->size(i);
+	if(child)
+		child->size(i);
+
+	return i;
+}
+
+
+template <typename T>
+void binomialNode<T>::addChild(binomialNode<T>* childNode)
+{
+	if(childNode)
+	{
+		childNode->sibling = 0;
+		if(child)
+		{
+			binomialNode<T>* temp = child;
+			while(temp->sibling)
+				temp = temp->sibling;
+			temp->sibling = childNode;
+		}
+		else
+		{
+			child = childNode;
+		}
+		//
+		order++;
+		childNode->parent = this;
+	}
+}
+template <typename T>
+binomialNode<T>* binomialNode<T>::find(T v)
+{
+	binomialNode<T>* found = 0;
+	if(v == value)
+		return this;
+	else if(sibling)
+	{
+		found =  sibling->find(v);
+	}
+	if( value < v && ! found) //if v is larger, check children,
+	{
+		if(child)
+		{
+			found =  child->find(v);
+		}
+	}
+
+	return found;
+}
+template <typename T>
+void binomialNode<T>::print(std::ostream& out)
+{
+	out << "\nNode : " << std::setw(5) << value;
+	if(DEBUG)
+	{
+		out << "\tOrder:  " << order;
+		out << "\tParent  : " ;
+		if(parent) out  << std::setw(5) << parent ->value;
+		out << "\tSibling : " ;
+		if(sibling) out  << std::setw(5) << sibling->value;
+		out << "\tChild   : " ;
+		if(child) out << std::setw(5) << child->value;
+		if(sibling && sibling->order < order)
+			out << "********ERROR IN SIBLING ORDERING!!!******";
+	}
+	if(sibling)
+		sibling->print(out);
+	if(child)
+			child->print(out);
+
+}
+
+template <typename T>
+class binomialHeap
+{
+	friend class binomialNode<T>;
+
+public:
+	binomialHeap();
+	~binomialHeap();
+	/*
+	 * insert
+	 * 		Creates a singleton heap node with value v and
+	 * 		 unions it with the head
+	 * 	value:
+	 * 		value to insert into heap
+	 *
+	 */
+	void insert(T value);
+	/*
+	 * getMin
+	 * 		Searches the heap list for the node with minimum value
+	 *
+	 * returns
+	 * 		minimum value in heap list
+	 */
+	T getMin();
+	/*
+	 * extractMin
+	 * 		Removes the heap list for node with minimum value
+	 * 		and unions minimum node's children with heap list
+	 *
+	 * returns
+	 * 		minimum value in heap list
+	 */
+	T extractMin();
+	/*
+	 * deleteKey
+	 * 		Removes the node with the value v by decreasing
+	 * 		node value to one less than current min and then
+	 * 		extracts the min.
+	 * value:
+	 * 		value to delete from heap
+	 *
+	 * returns
+	 * 		true if successfully deleted
+	 */
+	bool deleteKey(T value);
+	/*
+	 * decreaseKey
+	 * 		Searches for node with value v a
+	 *
+	 * 	value:
+	 * 		value to find and decrease
+	 * 	newValue:
+	 * 		value to decrease value to
+	 * returns
+	 * 		true if successfully decreased
+	 */
+	bool decreaseKey(T value, T newValue);
+	/*
+	 * size
+	 * 		Number of nodes in heap.
+	 *
+	 * returns
+	 *		number of nodes in heap
+	 */
+	int size();
+	bool isEmpty();
+	void print(std::ostream& out);
```

Comment:
Same goes as above. Here I would ditch the idea of a print function, and instad implement iterators so that a user could iterate over the heap and handle output themselves.